src/Reflection/ClassReflection.php

```diff
@@ -489,6 +588,76 @@ public function isFinal(): bool
 		return $this->isFinal;
 	}
 
+	public function getTemplateTypeMap(): TemplateTypeMap
+	{
+		if ($this->templateTypeMap !== null) {
+			return $this->templateTypeMap;
+		}
+
+		$resolvedPhpDoc = $this->getResolvedPhpDoc();
+		if ($resolvedPhpDoc === null) {
+			$this->templateTypeMap = TemplateTypeMap::createEmpty();
+			return $this->templateTypeMap;
+		}
+
+		$templateTypeMap = new TemplateTypeMap(array_map(function (TemplateTag $tag): Type {
+			return TemplateTypeFactory::fromTemplateTag(
+				TemplateTypeScope::createWithClass($this->getName()),
+				$tag
+			);
+		}, $this->getTemplateTags()));
+
+		$this->templateTypeMap = $templateTypeMap;
+
+		return $templateTypeMap;
+	}
+
+	public function getResolvedTemplateTypeMap(): ?TemplateTypeMap
+	{
+		return $this->resolvedTemplateTypeMap;
+	}
+
+	public function getActiveTemplateTypeMap(): TemplateTypeMap
+	{
+		return $this->resolvedTemplateTypeMap ?? $this->getTemplateTypeMap();
+	}
+
+	public function isGeneric(): bool
+	{
+		return count($this->getTemplateTags()) > 0;
+	}
+
+	public function typeMapFromList(array $types): TemplateTypeMap
+	{
+		$resolvedPhpDoc = $this->getResolvedPhpDoc();
+		if ($resolvedPhpDoc === null) {
+			return TemplateTypeMap::createEmpty();
+		}
+
+		$map = [];
+		$i = 0;
+		foreach ($resolvedPhpDoc->getTemplateTags() as $tag) {
+			$map[$tag->getName()] = $types[$i] ?? new ErrorType();
+			$i++;
+		}
+
+		return new TemplateTypeMap($map);
+	}
+
+	public function withTypes(array $types): self
```

Comment:
It needs to be specified what's in the array here.

Reply:
(There's a rule for this in slevomat/coding-standard but we'd need 5.x so that `array<x, y>` syntax is understood, but 5.x requires PHP 7.2 which we can't yet use in our build. There's no many reasons to bump minimum PHP version to 7.2 so I'm postponing it.)